Source/JavaScriptCore/dfg/DFGFixupPhase.cpp

```diff
@@ -4117,13 +4117,39 @@ class FixupPhase : public Phase {
                 attemptToForceStringArrayModeByToStringConversion<StringOrStringObjectUse>(arrayMode, node);
         }
             
-        if (!arrayMode.supportsSelfLength())
+        if (!arrayMode.supportsSelfLength()) {
+            if (arrayMode.type() == Array::AlwaysSlowPutContiguous)
+                return attemptToMakeGetArrayLengthForAlwaysSlowPutContiguous(node, arrayMode);
+
             return false;
+        }
         
         convertToGetArrayLength(node, arrayMode);
         return true;
     }
 
+    bool attemptToMakeGetArrayLengthForAlwaysSlowPutContiguous(Node* node, ArrayMode arrayMode)
+    {
+        if (arrayMode.arrayClass() != Array::OriginalNonArray)
+            return false;
+
+        JSGlobalObject* globalObject = m_graph.globalObjectFor(node->origin.semantic);
+        Structure* objectPrototypeStructure = globalObject->objectPrototype()->structure();
+
+        if (globalObject->alwaysSlowPutContiguousWatchpointSet().isStillValid()
+            && objectPrototypeStructure->transitionWatchpointSetIsStillValid()
+            && globalObject->objectPrototypeIsSaneConcurrently(objectPrototypeStructure)) {
+
+            m_graph.watchpoints().addLazily(globalObject->alwaysSlowPutContiguousWatchpointSet());
+            m_graph.registerAndWatchStructureTransition(objectPrototypeStructure);
+
+            convertToGetArrayLength(node, arrayMode);
+            return true;
+        }
+
+        return false;
+    }
```

Comment:
We should only do this if profiling tells us we're the original structure, right? Otherwise, this is an OSR exit loop AFAICT

Reply:
Ignore me, I see that we're relying on profiling above that we're "original"